File: Dynamic Programming/DP on 1D/Frog Jump with k distances/Tabulation.java
// User function Template for Java

class Solution {
    public int minimizeCost(int k, int arr[]) {
        int n = arr.length;
        int[] dp = new int[n];
        
        dp[0] = 0;
        
        for (int i = 1; i < n; i++) {
            int minCost = Integer.MAX_VALUE;
            
            for (int j = 1; j <= k; j++) {
                if (i - j >= 0) {
                    int cost = Math.abs(arr[i] - arr[i - j]) + dp[i - j];
                    minCost = Math.min(minCost, cost);
                }
            }
            
            dp[i] = minCost;  
        }
        
        return dp[n - 1];  
    }
}

File: Dynamic Programming/DP on MCM - Partition DP/Matrix Chain Multiplication/tabulation.cpp
int matrixMultiplication(vector<int> &arr) {
    int n = arr.size();
    int dp[n][n];
    
    for(int i=1; i<n; i++) dp[i][i] = 0;
    
    for(int i=n-1; i>=1; i--){
        for(int j=i+1; j<=n-1; j++){
            int mini = 1e9;
            for(int k=i; k<=j-1; k++){
                int temp = arr[i-1]*arr[k]*arr[j] + dp[i][k] + dp[k+1][j];
                mini = min(mini, temp);
            }
            dp[i][j] = mini;
        }
    }
    
    return dp[1][n-1];
}